test/integration/crl_test.go

```diff
@@ -97,3 +101,72 @@ func TestCRLPipeline(t *testing.T) {
 	test.AssertEquals(t, string(reason), "5")
 	resp.Body.Close()
 }
+
+func TestTemporalAndExplicitShardingCoexist(t *testing.T) {
+	db, err := sql.Open("mysql", vars.DBConnSAIntegrationFullPerms)
+	test.AssertNotError(t, err, "opening database connection")
+	// Insert an old, revoked certificate in the certificateStatus table. Importantly this
+	// serial has the 7f prefix, which is in test/config-next/crl-updater.json in the
+	// `temporallyShardedPrefixes` list.
+	oldSerial := "7faa39be44fc95f3d19befe3cb715848e601"
+	issuerID := 43104258997432926
+	_, err = db.Query(`DELETE FROM certificateStatus WHERE serial = ?`, oldSerial)
+	if err != nil {
+		t.Fatalf("deleting old certificateStatus row: %s", err)
+	}
+	_, err = db.Exec(`DELETE FROM certificateStatus WHERE serial = ?`, oldSerial)
+	if err != nil {
+		t.Fatalf("deleting old certificateStatus row: %s", err)
+	}
+	_, err = db.Exec(fmt.Sprintf(`
+		INSERT INTO certificateStatus (serial, issuerID, notAfter, status, ocspLastUpdated, revokedDate, revokedReason, lastExpirationNagSent)
+		VALUES ("%s", %d, "%s", "revoked", NOW(), NOW(), 0, 0);`,
+		oldSerial, issuerID, time.Now().Add(24*time.Hour).Format("2006-01-02 15:04:05")))
+	test.AssertNotError(t, err, "inserting old revoked certificate")
+
+	client, err := makeClient()
+	test.AssertNotError(t, err, "creating acme client")
+
+	certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	test.AssertNotError(t, err, "creating random cert key")
+
+	domain := random_domain()
+	result, err := authAndIssue(client, certKey, []string{domain}, true)
+	if err != nil {
+		t.Fatalf("authAndIssue: %s", err)
+	}
+	err = client.RevokeCertificate(
+		client.Account,
+		result.certs[0],
+		client.PrivateKey,
+		0,
+	)
+	test.AssertNotError(t, err, "revocation should have succeeded")
+
+	// Reset the "leasedUntil" column to prepare for another round of CRLs.
+	fc := clock.NewFake()
+	_, err = db.Exec(`UPDATE crlShards SET leasedUntil = ?`, fc.Now().Add(-time.Minute))
+	test.AssertNotError(t, err, "resetting leasedUntil column")
+
+	runUpdater(t, path.Join(os.Getenv("BOULDER_CONFIG_DIR"), "crl-updater.json"))
+
+	allCRLs := getAllCRLs(t)
+	allSeen := make(map[string]bool)
+	for _, crls := range allCRLs {
+		seen := make(map[string]bool)
```

Comment:
Three small comments:
1) Rather than looping over allCRLs, can't you just index into it since you know which issuer issued this cert?
2) If not, can't you just use a single `seen` map, since if this cert appears on multiple different issuer's CRLs that's a problem too?
3) You only reference allCRLs once, so you can just inline the function call

All together that would lead to:
```suggestion
	seen := make(map[string]bool)
	for _, crls := range getAllCRLs(t) {
```

Reply:
I remembered why I did it that way. The fake "old" certificate that we insert into the DB has a different issuer than the more realistic "new" certificate that we generate `authAndIssue`. And we don't actually have the AKID for the "old" certificate.

Reply:
> why does it have two wholly different ways of putting entries in the database?
> For the "old" cert, you directly manipulate the certificateStatus table. This I think makes sense, because it means you're manually controlling the serial to ensure it has the "old" prefix.
> For the "new" cert, you're doing a whole issuance flow.
> Why can't you directly manipulate the certificateStatus table (and the revokedCertificates table, I guess?) for both?

In general I have a strong preference for the integration tests to use the ACME issuance flow. That's what makes them integration tests, and it's more realistic. But for the "old" cert, that serial prefix no longer exists in the configs, so we can't issue for it through the ACME interface.

I did try having two CAs in config-next, one with the "old" prefix (and no IncludeCRLDistributionPoints) and one with the "new" prefix, but that didn't work out. Issuances failed because the precert would hit one CA and the final cert would hit a different one.

Reply:
Fair enough. I still think this test needs some added clarity, though. I've added some notes above that I think would have helped me.